newrelic/config.py

```diff
@@ -3169,6 +3176,23 @@ def _setup_agent_console():
         newrelic.core.agent.Agent.run_on_startup(_startup_agent_console)
 
 
+def _setup_security_module():
+    """Initiates k2 security module and adds a
+    callback to agent startup to propagate NR config
+    """
+    try:
+        if not _settings.security.agent.enabled:
+            return
+        from newrelic_security.api.agent import Agent as SecurityAgent
+
+        # initialize security agent
+        security_agent = SecurityAgent()
+        # create a callback to reinitialise the security module
+        newrelic.core.agent.Agent.run_on_startup(security_agent.refresh_agent)
+    except Exception as k2error:
+        _logger.error("K2 Startup failed with error %s", k2error)
+
```

Comment:
```suggestion
def _setup_security_module():
    def _security_agent_registration_callback():
        """Callback that initializes k2 security module and propagates NR config."""
        try:
            from newrelic.api.application import application_instance
            application = application_instance(activate=False)
            if not application or not application.settings or not application.settings.security.agent.enabled:
                return
            
            from newrelic_security.api.agent import Agent as SecurityAgent
            
            # initialize security agent
            security_agent = SecurityAgent()
            security_agent.refresh_agent()

        except Exception as k2error:
            _logger.error("K2 Startup failed with error %s", k2error)

    # create a callback to only initialize the security module after high security settings have been applied
    newrelic.core.agent.Agent.run_on_registration(_security_agent_registration_callback)
```

Reply:
For compliance I think we can't run any of this until high security mode has a chance to disable it. We'll need to wait until after connect and registration, and also use application instance settings instead of the global settings (which don't include server side fixups like high security).

I think this callback should do the trick, but that depends on if

1. Your agent is a singleton and can be called like this multiple times with no issues.
2. Your agent can handle being called this late into the import process.

Reply:
We should DEFINITELY add tests for this, somehow.

Reply:
The above code suggestion will also need the application instance passed to the call to `run_on_registration()`